### ISSUE TITLE
Descriptor Loss and SuperPoint training

### ISSUE THREAD

**Issue by ppinheiro2** [rpautrat/SuperPoint#157]
Hi @rpautrat, I'm trying to implement SuperPoint paper in PyTorch and your repository is helping me a lot, thank you. I'm having some difficulties training the descriptors in SuperPoint.
The detections seem to be good as I've evaluated the repeatability and the values are similar to yours and MagicLeap pre-trained weights. However, when I try to match the descriptors in HPatches images I get really less matches than with MagicLeap pre-trained weights.
I've tried several values for lambda_d and lambda_loss, could you explain me why yours are so different than those in the paper? And how did you get to them?
Could you show me the values of your descriptor loss during training? I would like to understand if my descriptor loss is well written.

Thank you for your work

**Comment by rpautrat** [comment]
Hi @ppinheiro2, I am glad to hear that this repo is helping you for your own implementation.

In the last version of this repo, I actually slightly modified the original loss of the descriptor. The loss is now computed on L2 normalized descriptors, and after computing the distance matrix between the descriptors of both images, I apply ReLU to eliminate obvious non matches (similarity of the descriptor below 0) as described in this [paper](https://arxiv.org/pdf/1703.05593.pdf) (see page 3, paragraph 'Matching layer'), and I renormalize the distance matrix to help disambiguate when several descriptors are very close. You can find these modifications in the [descriptor loss computation](https://github.com/rpautrat/SuperPoint/blob/61bc16ff0e096340a26feb4cf80bf6b4bacee267/superpoint/models/utils.py#L110) This was not part of the original paper, but I found empirically that it improved the convergence and the results. This explains why I have lambda_loss so different, as the new loss has a different magnitude as the original one.

As for the different lambda_d, since the distance matrix has been normalized, the positive matches have a higher activation than all the negative ones, so I had to reduce lambda_d to balance the negative and positive matches.

These two values have been found empirically by running a grid search over the parameters.

During training, my custom descriptor loss is around 0.05 (including the multiplication by lambda_loss) after convergence, for a total loss of around 1.

**Comment by ppinheiro2** [comment]
Thank you very much for your detailed answer. 

I think I understood the modifications now. I am training my model again to see if the results are what I expected. Do you train for 600000 iterations? Your loss converges to 1 after that? And you only save the model weights after that last iteration?

**Comment by rpautrat** [comment]
I actually trained for 300 000 iterations in my last training. The loss had converged to 1 at that point, and the model available online corresponds to the weights of that iteration, yes.

**Comment by ppinheiro2** [comment]
I left my model training and now I'm seeing the results. 
These are my qualitative results on HPatches with 1000 keypoints in each image
![v_yard_matches](https://user-images.githubusercontent.com/68388976/88071001-39ce9300-cb6b-11ea-84bd-97e1b8c76999.png)
![i_crownnight_matches](https://user-images.githubusercontent.com/68388976/88071049-47841880-cb6b-11ea-8f19-c7dfe47ac91f.png)
Repeatability: Viewpoint-> 0.539 Illumination->0.671
Homography estimation on HPatches computed with a maximum of 1000 points detected in common between pairs of images, a threshold of correctness of 3 and with a NMS of 8: Viewpoint-> 0.742 Illumination-> 0.926

I have some questions,
1) My descriptor loss converges to about 0.78 and the total loss to about 1.55, which are different values than yours. Do you know what could be causing that?
2) Qualitative results are similar to magicleap pretrained weights and the quantitative results are similar too, except for the case of the viewpoint repeatability.
3) I've been saving the weights on different epochs and after some iterations something strange begins happen sometimes with grid like detections
![i_crownnight_matches_bad](https://user-images.githubusercontent.com/68388976/88073187-d09c4f00-cb6d-11ea-838b-286e368346da.png)

If you prefer, you could send me an e-mail to pedro.maiapinheiro2@gmail.com so we could discuss this questions.
I greatly appreciate your help!

**Comment by rpautrat** [comment]
Hi, yes let's discuss this per email. I am closing the issue then.

**Comment by shaofengzeng** [comment]
@ppinheiro2 Have you solve this problem? I also have the same problem,  hope you can help me

**Comment by shaofengzeng** [comment]
@rpautrat @ppinheiro2 Excese me but i really hope to get your help, i also have the same problem, have you find a solution ?

**Comment by shaofengzeng** [comment]
@rpautrat @ppinheiro2 Excese me but i really hope to get your help, i also have the same problem, have you find a solution ?
